### ISSUE TITLE
time to do memory minimization increases with each round

### ISSUE THREAD

**Issue by mykmelez** [mozilla/j2me.js#1508]
When running the startup benchmark on device, the time to do memory minimization increases with each round, until it takes a minute or longer in the final rounds, which makes it prohibitive to run the benchmark on device. The problem is intermittent, although right now I can reproduce consistently on both 2.0 and 2.1 Flames. I've also seen it on desktop, although rarely.

**Comment by mykmelez** [comment]
Accompanying this issue is a large delta between min and max startup times (f.e. 20 - 46 seconds).

**Comment by mykmelez** [comment]
I tried changing to a user build of FxOS, since @marco-c and @brendandahl both use them. To root it, so I could set the pref that enables memory minimization, I followed the steps in https://wiki.mozilla.org/B2G/QA/Tips_And_Tricks#Rooting_a_production_device, using a boot.img file I got from an engineer build on pvtbuilds.mozilla.org. But the benchmark exhibited the same problem.

**Comment by mykmelez** [comment]
@marco-c sets *config.autosize* to `"0"` and runs the benchmark by pressing the button in the debug console, so I tried doing that, but it exhibited the same problem.

**Comment by mykmelez** [comment]
FWIW, I don't see this problem when I test a midlet by starting to its initial screen (the one you see when you aren't registered). I only see it when starting to the screen it shows to registered users. And @brendandahl tests the midlet by starting to its initial screen, which may explain why he doesn't see it either.

But I can't explain why @marco-c doesn't see it, since he tests by starting to the registered screen.

**Comment by mykmelez** [comment]
When I run `tools/get_about_memory.py --minimize` while the startup benchmark is running, then the app doesn't show up in the memory report, which @EricRahm suggests is because the profiler's request timed out before the app responded to it.

When I run `tools/get_about_memory.py` (without ` --minimize`) while the startup benchmark is running, then the app shows up in the memory report, and it reports a very large detached window, f.e.:

│  ├──257.13 MB (70.61%) -- top(none)/detached/window(app://DOMAIN/main.html)

**Comment by mykmelez** [comment]
That detached window tends to grow over time as the benchmark run continues:

│  ├──366.90 MB (71.58%) ++ top(none)/detached/window(app://DOMAIN/main.html)

That's strange. We do reload the page, which presumably detaches a window (and eventually should GC it). But we reload the page over and over again. If those pages' window objects were leaking, I'd expect to see multiple entries for detached windows, one for each page we loaded, rather than one detached window that grows over time.

**Comment by mykmelez** [comment]
NB: if you measure memory enough, you may eventually run out of space to store the reports in /data/local/tmp/memory-reports/, which may manifest as a `DOMException [NS_ERROR_DOM_QUOTA_REACHED: "Persistent storage maximum size reached"` or `QuotaExceededError`. You can fix the problem by removing those reports via `adb shell`.

**Comment by mykmelez** [comment]
Strangely, the detached window does *not* show up in a memory report after I disable *security.turn_off_all_security_so_that_viruses_can_take_over_this_computer*, so the benchmark doesn't itself minimize/measure memory. Perhaps the benchmark code for introspecting its own process's memory usage causes the window to leak?

**Comment by mykmelez** [comment]
I beat my head against this problem some more today, and I discovered that I could manually reproduce it by exposing *Benchmark.sampleMemoryToStorage* and calling it after a midlet loads. But the behavior was timing-dependent: if I waited 10-15 seconds after the midlet loaded, I didn't see it. If I called it a few seconds after the midlet loaded, however, then I did.

So I tried changing *roundDelay* from 5000 to 10000ms and running the benchmark with 30 rounds, after which I still saw the problem, but it seemed like less memory had been leaked overall. So perhaps it's specific to something the midlet or the VM does shortly after startup that takes a variable number of seconds to complete.

Finally, I updated a Flame to the latest Central nightly build of B2G and tried to reproduce the problem there. Without memory minimization, I saw six "ghost" (i.e. "detached") windows after a benchmark run of 30 rounds with a 5000ms delay. But minimization via B2G's get_about_memory.py reaped those ghosts. And a similar benchmark run with minimization between rounds never exhibited the problem.

So it seems to be a JS engine bug, and it seems to be fixed on trunk. Thus I'm going to close this issue. If you're experiencing it, the workaround is to use a trunk build, at least for startup benchmarking.
